File: ios/Classes/NavigineRouteSession.swift
import Foundation
import Navigine

public class NavigineRouteSession: NSObject, NCAsyncRouteListener {
    private var id: Int
    private var session: NCRouteSession
    private let methodChannel: FlutterMethodChannel!

    private var currentRoutePath: NCRoutePath?

    public required init(
        id: Int,
        session: NCRouteSession,
        registrar: FlutterPluginRegistrar
    ) {
        self.id = id
        self.session = session

        methodChannel = FlutterMethodChannel(
            name: "navigine_sdk/navigine_route_session_\(id)",
            binaryMessenger: registrar.messenger()
        )

        super.init()
        weak var weakSelf = self
        self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })

        self.session.add(weakSelf)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "split":
            result(split(call))
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func split(_ call: FlutterMethodCall) -> [[String: Any?]] {
        let params = call.arguments as! [String: Any]
        let distance = params["distance"] as! NSNumber

        var paths = [[String: Any?]]()
        if currentRoutePath != nil {
            for path in currentRoutePath!.split(distance.floatValue) {
                paths.append(Utils.routePathToJson(path))
            }
        }

        return paths
    }

    public func unsubscribe() {
        self.session.remove(self)
    }


    public func getSession() -> NCRouteSession {
        return self.session
    }

    public func onRouteChanged(_ currentPath: NCRoutePath?) {
        var arguments: [String: Any?] = [:]
        if (currentPath != nil) {
             arguments = [
                "routePath": Utils.routePathToJson(currentPath!)
            ]
        }
        currentRoutePath = currentPath
        methodChannel.invokeMethod("onRouteChanged", arguments: arguments)
    }

    public func onRouteAdvanced(_ distance: Float, point: NCLocationPoint) {
        let arguments = [
            "distance": distance,
            "locationPoint": Utils.locationPointToJson(point)
        ] as [String : Any]
        methodChannel.invokeMethod("onRouteAdvanced", arguments: arguments)
    }
}

File: android/src/main/java/com/navigine/naviginesdk/NavigineRouteSession.java
package com.navigine.naviginesdk;

import androidx.annotation.NonNull;

import com.navigine.idl.java.AsyncRouteListener;
import com.navigine.idl.java.LocationPoint;
import com.navigine.idl.java.RoutePath;
import com.navigine.idl.java.RouteSession;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class NavigineRouteSession implements MethodChannel.MethodCallHandler {
  private final int id;
  private final RouteSession session;
  private final MethodChannel methodChannel;

  private RoutePath currentRoutePath = null;
  private AsyncRouteListener asyncRouteListener = null;

  public NavigineRouteSession(
    int id,
    RouteSession session,
    BinaryMessenger messenger
  ) {
    this.id = id;
    this.session = session;

    methodChannel = new MethodChannel(messenger, "navigine_sdk/navigine_route_session_" + id);
    methodChannel.setMethodCallHandler(this);

    asyncRouteListener = new AsyncRouteListener() {
      @Override
      public void onRouteChanged(RoutePath routePath) {
        Map<String, Object> arguments = new HashMap<>();
        if (routePath != null) {
          arguments.put("routePath", Utils.routePathToJson(routePath));
        }
        currentRoutePath = routePath;
        methodChannel.invokeMethod("onRouteChanged", arguments);
      }

      @Override
      public void onRouteAdvanced(float v, LocationPoint locationPoint) {
        Map<String, Object> arguments = new HashMap<>();
        arguments.put("distance", v);
        arguments.put("locationPoint", Utils.locationPointToJson(locationPoint));
        methodChannel.invokeMethod("onRouteAdvanced", arguments);
      }
    };

    this.session.addRouteListener(asyncRouteListener);
  }

  public RouteSession getSession() {
    return this.session;
  }

  public void unsubscribe() {
    this.session.removeRouteListener(asyncRouteListener);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {

      case "split":
        result.success(split(call));
        break;

      default:
        result.notImplemented();
        break;
    }
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public ArrayList<Map<String, Object>> split(MethodCall call) {
    Map<String, Object> params = (Map<String, Object>) call.arguments;
    Double distance = (Double) params.get("distance");

    ArrayList<Map<String, Object>> paths = new ArrayList<>();
    if (currentRoutePath != null) {
      for (RoutePath path : currentRoutePath.split(distance.floatValue())) {
        paths.add(Utils.routePathToJson(path));
      }
    }

    return paths;
  }
}
